File: core/src/main/java/com/gmail/marcosav2010/connection/NetworkIdentificator.java
package com.gmail.marcosav2010.connection;

import com.gmail.marcosav2010.peer.NetworkPeer;

import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

/**
 * This clase manages all remote peer connections to connected peers
 *
 * @author Marcos
 */
public class NetworkIdentificator<T extends NetworkPeer> {

    private final Map<UUID, T> connectionPeer;
    private final Map<UUID, UUID> connectionPeerUUID;

    public NetworkIdentificator() {
        connectionPeer = new ConcurrentHashMap<>();
        connectionPeerUUID = new ConcurrentHashMap<>();
    }

    public T getPeer(UUID uuid) {
        return connectionPeer.get(uuid);
    }

    public UUID getConnection(NetworkPeer peer) {
        return connectionPeer.entrySet().stream().filter(e -> e.getValue().equals(peer)).map(Map.Entry::getKey).findFirst().orElseGet(null);
    }

    public boolean hasPeer(UUID uuid) {
        return connectionPeer.containsKey(uuid);
    }

    public boolean hasPeer(NetworkPeer peer) {
        return connectionPeerUUID.containsKey(peer.getUUID());
    }

    protected void put(UUID connectionUUID, T peer) {
        connectionPeerUUID.put(connectionUUID, peer.getUUID());
        connectionPeer.put(connectionUUID, peer);
    }

    protected T remove(UUID connectionUUID) {
        connectionPeerUUID.remove(connectionUUID);
        return connectionPeer.remove(connectionUUID);
    }

    public Set<T> getConnectedPeers() {
        return Set.copyOf(connectionPeer.values());
    }

    public int size() {
        return connectionPeer.size();
    }
}

File: core/src/main/java/com/gmail/marcosav2010/connection/NetworkIdentificator.kt
package com.gmail.marcosav2010.connection

import com.gmail.marcosav2010.peer.NetworkPeer
import java.util.*
import java.util.concurrent.ConcurrentHashMap

/**
 * This class manages all remote peer connections to connected peers
 *
 * @author Marcos
 */
open class NetworkIdentificator<T : NetworkPeer> {

    private val connectionPeer: MutableMap<UUID, T> = ConcurrentHashMap()
    private val connectionPeerUUID: MutableMap<UUID, UUID> = ConcurrentHashMap()

    fun getPeer(uuid: UUID): T? = connectionPeer[uuid]

    fun getConnection(peer: NetworkPeer) =
            connectionPeer.entries.filter { e -> e.value == peer }.map { it.key }.firstOrNull()

    fun hasPeer(uuid: UUID) = connectionPeer.containsKey(uuid)

    fun hasPeer(peer: NetworkPeer) = connectionPeerUUID.containsKey(peer.uuid)

    protected fun put(connectionUUID: UUID, peer: T) {
        connectionPeerUUID[connectionUUID] = peer.uuid
        connectionPeer[connectionUUID] = peer
    }

    protected fun remove(connectionUUID: UUID): T? {
        connectionPeerUUID.remove(connectionUUID)
        return connectionPeer.remove(connectionUUID)
    }

    val connectedPeers: Set<T>
        get() = connectionPeer.values.toSet()

    fun size() = connectionPeer.size
}